extensions/lifecycle/tasks/LifecycleTask.js

```diff
@@ -190,55 +205,131 @@ class LifecycleTask extends BackbeatTask {
     }
 
     /**
-     * For all incomplete MPU uploads, compare with rules (based only on prefix
-     * because no tags exist for incomplete MPU), and apply
-     * AbortIncompleteMultipartUpload rule (if any) on each upload
+     * Helper method to merge and sort Versions and DeleteMarkers lists
+     * @param {array} versions - versions list
+     * @param {array} deleteMarkers - delete markers list
+     * @return {array} merge sorted array
+     */
+    _mergeSortVersionsLists(versions, deleteMarkers) {
+        const sortedList = [];
+        let vIdx = 0;
+        let dmIdx = 0;
+
+        while (vIdx < versions.length || dmIdx < deleteMarkers.length) {
+            if (versions[vIdx] === undefined) {
+                sortedList.push(deleteMarkers[dmIdx++]);
+            } else if (deleteMarkers[dmIdx] === undefined) {
+                sortedList.push(versions[vIdx++]);
+            } else if (versions[vIdx].Key !== deleteMarkers[dmIdx].Key) {
+                // 1. by Key name
+                const isVKeyEarlier = (versions[vIdx].Key <
+                    deleteMarkers[dmIdx].Key);
+                if (isVKeyEarlier) {
+                    sortedList.push(versions[vIdx++]);
+                } else {
+                    sortedList.push(deleteMarkers[dmIdx++]);
+                }
+            } else if (versions[vIdx].LastModified !==
```

Comment:
To precise my thoughts, you still need to consider `isLatest` so that you don't expire current versions with the `NoncurrentVersionExpiration` rule, but it should not be necessary to make it a sorting criteria.